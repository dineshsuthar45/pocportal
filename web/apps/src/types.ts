import { InputProps, TypographyProps } from "@mui/material";
import { CSSProperties, ChangeEvent } from "react";

export interface IThemeSchema {
  palette: any;
  typography: any;
  components?: any;
}

export interface IGlobalTableProps {
  row: any;
  list: any;
}

export interface IGlobalButtonProps {
  label: any;
  sublabel?: string;
  disabled?: boolean;
  onClick?: () => void;
  style?: CSSProperties;
  type?: string;
}

export interface IGlobalImgProps {
  img: string;
  alt: string;
  style?: CSSProperties;
}

export interface IGlobaltextFieldProps {
  id?: string;
  value: string;
  label: string;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
  name: string;
  errmsg?: string;
  type?: string;
  disabled?: boolean;
  style?: CSSProperties;
  placeholder?: string;
  showPassword?: boolean;
  InputProps?: InputProps;
}

export interface IGlobalTypoProps {
  id?: number;
  variant: TypographyProps["variant"];
  label: string;
  sublabel?: string;
  style?: CSSProperties;
}
