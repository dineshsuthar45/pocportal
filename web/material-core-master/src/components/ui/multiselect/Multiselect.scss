.MuiFormControlLabel-root {
	width: 100%;

	&.MuiFormControlLabel-labelPlacementTop {
		align-items: flex-start;
	}
}

.multiselect-root {
	width: 100%;

	.hide {
		display: none;
	}

	.m-select {
		cursor: pointer;
		position: relative;

		.selected {
			position: relative;
			padding: 4px 4px 4px 0px;
			display: flex;

			&::before {
				content: '';
				position: absolute;
				bottom: 0px;
				width: 100%;
				border-bottom: 1px solid transparent;
			}

			.selected-text {
				flex-grow: 1;
				text-overflow: ellipsis;
				white-space: nowrap;
				overflow: hidden;
			}
			.chevron {
				margin-left: auto;
			}
		}

		.options {
			position: absolute;
			background-color: white;
			opacity: 1;
			z-index: 1;
			width: 100%;
			display: flex;
			flex-direction: column;
			box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);

			&.scroll {
				overflow: auto;
			}

			.search {
				padding: 1px 2px;
			}

			.option {
				cursor: pointer;

				.MuiCheckbox-root {
					padding: 4px;
				}
			}

			.MuiFormControlLabel-root {
				margin: 0px;

				input[type='checkbox'] {
					margin-left: 0px;
				}

				.MuiFormControlLabel-label {
					padding-left: 4px;
				}
			}

			.no-options {
				height: 40px;
				display: flex;
				align-items: center;
				justify-content: center;
			}
		}
	}
}
