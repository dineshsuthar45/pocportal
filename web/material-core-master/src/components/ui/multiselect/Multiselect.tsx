import { useTheme } from '@mui/material';
import classNames from 'classnames';
import { useEffect, useMemo, useRef, useState } from 'react';
import { useClickOutsideMultipleRefs } from '../../../hooks';
import { MCheckbox } from '../checkbox/Checkbox';
import { MIconAngleDown, MIconAngleUp } from '../icons';
import { MFormControlLabel } from '../internal/formControlLabel';
import { MSearch } from '../search/Search';
import { MTooltip } from '../tooltip/Tooltip';
import './Multiselect.scss';
import { MultiselectProps } from './Multiselect.types';
import './MultiselectDark.scss';
import './MultiselectLight.scss';

/**
 * Multiselect component
 */
export const MMultiselect = <T extends { label: string; value: string; checked?: boolean }>({
	debounceOnSearch = 250,
	disabled = false,
	label,
	labelPlacement = 'top',
	optionsNotSupplied = 'No options supplied.',
	placeholder = 'Select',
	selectAll = false,
	checkboxSize = 'small',
	visibleItems = 10,
	...props
}: MultiselectProps<T>) => {
	const theme = useTheme();

	const optionsRef = useRef<HTMLDivElement>(null);
	const selectedTextRef = useRef<HTMLDivElement>(null);
	const MSearchRef = useRef<HTMLDivElement>(null);
	const [initializationDone, setInitializationDone] = useState(false);
	const [isSelectAllIndeterminate, setSelectAllIndeterminate] = useState(false);
	const [isSelectAll, setIsSelectAll] = useState(selectAll);
	const [isMenuOpen, setIsMenuOpen] = useState(false);
	const [searchText, setSearchText] = useState('');
	const [selectedText, setSelectedText] = useState('');
	const [options, setOptions] = useState<T[]>(props.options);
	const [displayOptions, setDisplayOptions] = useState<JSX.Element[]>([<></>]);
	const [themeMode, setThemeMode] = useState<'light' | 'dark'>(theme?.palette?.mode ?? 'light');

	useEffect(() => {
		setThemeMode(theme.palette.mode);
	}, [theme.palette.mode]);

	const doesItFit = (newTextContent: string): boolean => {
		if (selectedTextRef && selectedTextRef.current && selectedTextRef.current.parentElement) {
			const selectCloned = selectedTextRef.current.cloneNode(true);

			if (selectCloned.firstChild && selectCloned.firstChild.firstChild) {
				(selectCloned as HTMLElement).style.visibility = 'hidden';
				selectCloned.firstChild.firstChild.textContent = newTextContent;
				selectedTextRef.current.parentElement.appendChild(selectCloned);
				const scrollingActive = isHorizontalScrollingActive(selectCloned.firstChild as HTMLElement);
				selectedTextRef.current.parentElement.removeChild(selectCloned);
				return !scrollingActive;
			}
		}

		return false;
	};

	const isHorizontalScrollingActive = (e: HTMLElement): boolean => {
		return e.offsetWidth < e.scrollWidth;
	};

	const selectedOptions = useMemo((): T[] => options.filter((option: T) => option.checked), [options]);

	const myElRef = useClickOutsideMultipleRefs<HTMLDivElement>([MSearchRef], () => setIsMenuOpen(false));

	const setAndSortOptions = (options: T[]): void => {
		if (options.find((option: T) => option.checked)) {
			setSelectAllIndeterminate(true);
		}

		setOptions(
			options
				.sort((a: T, b: T) => a.label.localeCompare(b.label))
				.sort((a: T, b: T) => (!(a.checked !== b.checked) ? 0 : a.checked ? -1 : 1))
		);
	};

	useEffect(() => {
		const allChecked = [...options]
			.sort((a: T, b: T) => a.label.localeCompare(b.label))
			.map((option) => ({ checked: false, ...option }));

		if (isSelectAll) {
			allChecked.forEach((option) => {
				option.checked = true;
			});
		}

		setAndSortOptions(allChecked);

		setInitializationDone(true);
	}, []);

	useEffect(() => {
		handleFilterOptions(searchText);

		if (props.onChange) {
			props.onChange(options);
		}
	}, [options]);

	useEffect(() => {
		setAndSortOptions(props.options);
	}, [props.options]);

	useEffect(() => {
		if (isMenuOpen && props.onOpen) {
			props.onOpen();
			MSearchRef.current!.getElementsByTagName('input')[0].focus();
		} else if (!isMenuOpen && props.onClose) {
			props.onClose();
			setSearchText('');
			handleFilterOptions('');
		}
	}, [isMenuOpen]);

	useEffect(() => {
		if (selectedOptions.length === options.length) {
			setSelectedText('All');
			setIsSelectAll(true);
			setSelectAllIndeterminate(false);
		} else if (selectedOptions.length === 0) {
			setIsSelectAll(false);
			setSelectedText('');
			setSelectAllIndeterminate(false);
		} else {
			const content = selectedOptions
				.map((option: T) => option.label)
				.join(', ')
				.trimEnd();

			if (doesItFit(content)) {
				setSelectedText(content);
			} else {
				setSelectedText(`(${selectedOptions.length}) ${content}`);
			}
		}
	}, [selectedOptions]);

	useEffect(() => {
		if (initializationDone) {
			handleSelectAllChange(selectAll);
		}
	}, [selectAll]);

	const handleOptionSelectedChange = (option: T) => {
		const newSelectedOptions = [...options];
		const optionIndexFound = newSelectedOptions.findIndex((x: T) => x.value === option.value);
		const newCheckedStatus = !newSelectedOptions[optionIndexFound].checked;
		newSelectedOptions[optionIndexFound].checked = newCheckedStatus;

		setAndSortOptions(newSelectedOptions);
	};

	const handleSelectAllChange = (newValue: boolean) => {
		setIsSelectAll(newValue);
		let allChecked = options.map((option) => ({ ...option, checked: false }));
		if (newValue) {
			allChecked = options.map((option) => ({ ...option, checked: true }));
		}

		setAndSortOptions(allChecked);
	};

	const handleSelectedClick = () => {
		setIsMenuOpen(!isMenuOpen);
	};

	const getOptionHeight = (): number => {
		const allOptions = optionsRef.current?.getElementsByClassName('option') ?? null;
		return allOptions && allOptions.length > 0 ? allOptions[0].clientHeight : 20;
	};

	const handleFilterOptions = (searchBy: string) => {
		const filteredOptions = [...options]
			.filter((option) => option.label.toLowerCase().includes(searchBy.toLowerCase()))
			.map((option: T, index: number) => {
				return (
					// todo Add disableInteractive to tooltip props
					<MCheckbox
						key={index}
						id={`option-${index}`}
						checked={option.checked}
						label={option.label}
						labelPlacement="end"
						size={checkboxSize}
						className="option"
						onChange={() => handleOptionSelectedChange(option)}
					/>
				);
			});
		setDisplayOptions(filteredOptions);
	};

	useEffect(() => {
		if (!props.onSearch) {
			handleFilterOptions(searchText);
		}
	}, [searchText, options]);

	useEffect(() => {
		if (isMenuOpen && optionsRef.current) {
			if (visibleItems * getOptionHeight() < (options.length + 2) * getOptionHeight()) {
				optionsRef.current.classList.add('scroll');
				optionsRef.current.style.height = `${visibleItems * getOptionHeight()}px`;
			} else {
				optionsRef.current.classList.remove('scroll');
				optionsRef.current.style.height = '';
			}
		}
	}, [visibleItems, options, isMenuOpen]);

	const noOptions = <div className="option no-options">{optionsNotSupplied}</div>;

	/* todo scrollbar component */
	const menuOptions = isMenuOpen && (
		<div ref={optionsRef} className="options" onScroll={(e) => props.onScroll && props.onScroll(e)}>
			{options.length <= 0 && noOptions}
			<div className={classNames('toolbar', { hide: options.length <= 0 })}>
				<div className="search">
					<MSearch
						ref={MSearchRef}
						debounceOnSearch={debounceOnSearch}
						placeholder="Search"
						value={searchText}
						onChange={(e) => setSearchText(e.target.value)}
						onReset={() => setSearchText('')}
						onSearch={(searchBy) => {
							if (props.onSearch) {
								props.onSearch(searchBy);
							}
						}}
					/>
				</div>
				<MCheckbox
					id="select-all"
					checked={isSelectAll}
					indeterminate={isSelectAllIndeterminate}
					label="All"
					labelPlacement="end"
					size={checkboxSize}
					className="option option-all"
					onChange={() => handleSelectAllChange(!isSelectAll)}
				/>
			</div>
			{displayOptions}
		</div>
	);

	const select = (
		<div className="m-select">
			<div
				ref={selectedTextRef}
				className={classNames('selected', {
					'menu-open': isMenuOpen,
					disabled,
				})}
				onClick={handleSelectedClick}
			>
				<MTooltip title={selectedText}>
					<span className="selected-text">
						{selectedOptions.length === 0 && (
							<span className={classNames('placeholder', { disabled })}>{placeholder}</span>
						)}
						{selectedOptions.length > 0 && selectedText}
					</span>
				</MTooltip>
				<div className={classNames('chevron', { disabled })}>
					<MIconAngleDown className={classNames({ hide: isMenuOpen })} />
					<MIconAngleUp className={classNames({ hide: !isMenuOpen })} />
				</div>
			</div>
			{menuOptions}
		</div>
	);

	return (
		<div
			ref={myElRef}
			className={classNames('multiselect-root', `m-theme-${themeMode}`, {
				disabled,
			})}
		>
			<MFormControlLabel
				label={label}
				labelPlacement={labelPlacement}
				tooltipProps={{
					title: 'internalTooltipTitle',
					showAlways: false,
				}}
				disabled={disabled}
			>
				{select}
			</MFormControlLabel>
		</div>
	);
};

MMultiselect.defaultProps = {
	debounceOnSearch: 250,
	disabled: false,
	labelPlacement: 'top',
	optionsNotSupplied: 'No options supplied.',
	placeholder: 'Select',
	selectAll: false,
	checkboxSize: 'small',
	visibleItems: 10,
};
