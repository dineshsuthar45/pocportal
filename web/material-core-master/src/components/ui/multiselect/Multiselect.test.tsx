import { render } from "@testing-library/react";
import React from "react";
import { MMultiselect } from "./Multiselect";
import { MultiselectProps } from "./Multiselect.types";

describe("Multiselect component", () => {
  const renderComponent = () => render(<MMultiselect {...props} />);
  let props: MultiselectProps<any>;

  beforeEach(() => {
    props = {
      label: "Some Multiselect text here",
      options: [
        {
          label: "Juan Francisco Fernandez Rodriguez",
          value: "1",
        },
        {
          label: "Dave",
          value: "2",
        },
        {
          label: "Chris",
          value: "3",
        },
        {
          label: "Rinat",
          value: "4",
        },
        {
          label: "Artemis",
          value: "5",
        },
        {
          label: "Natali",
          value: "6",
        },
        {
          label: "Surpreet",
          value: "7",
        },
        {
          label: "Sebastian",
          value: "8",
        },
      ],
    };
  });

  it("should render correctly", () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it("all options selected", () => {
    props.selectAll = true;

    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });
});
