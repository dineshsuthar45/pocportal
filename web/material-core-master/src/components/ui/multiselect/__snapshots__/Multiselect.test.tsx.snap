// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiselect component all options selected 1`] = `
<div>
  <div
    class="multiselect-root m-theme-light"
  >
    <div
      class="m-form-control-label top"
    >
      <div
        aria-label=""
        class="m-tooltip-root tooltip-label tooltip-container"
        data-mui-internal-clone-element="true"
      >
        <label>
          Some Multiselect text here
        </label>
      </div>
      <div
        class="label-children"
      >
        <div
          class="m-select"
        >
          <div
            class="selected"
          >
            <div
              aria-label=""
              class="m-tooltip-root tooltip-container"
              data-mui-internal-clone-element="true"
            >
              <span
                class="selected-text"
              >
                <span
                  class="placeholder"
                >
                  Select
                </span>
              </span>
            </div>
            <div
              class="chevron"
            >
              <svg
                class=""
                fill="#6F7880"
                height="10"
                viewBox="0 0 10 10"
                width="10"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m.5 3 4.1 4.1c.2.2.3.3.4.3s.2 0 .4-.2L9.6 3a.608.608 0 0 0 0-.8.608.608 0 0 0-.8 0L5 6.1 1.2 2.3a.483.483 0 0 0-.7 0 .483.483 0 0 0 0 .7Z"
                />
              </svg>
              <svg
                class="hide"
                fill="#6F7880"
                height="10"
                viewBox="0 0 10 10"
                width="10"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.6 6.5 5.4 2.3a.52.52 0 0 0-.4-.2.442.442 0 0 0-.4.2L.5 6.5a.385.385 0 0 0 0 .7.608.608 0 0 0 .8 0L5 3.4l3.8 3.8a.608.608 0 0 0 .8 0 .483.483 0 0 0 0-.7Z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Multiselect component should render correctly 1`] = `
<div>
  <div
    class="multiselect-root m-theme-light"
  >
    <div
      class="m-form-control-label top"
    >
      <div
        aria-label=""
        class="m-tooltip-root tooltip-label tooltip-container"
        data-mui-internal-clone-element="true"
      >
        <label>
          Some Multiselect text here
        </label>
      </div>
      <div
        class="label-children"
      >
        <div
          class="m-select"
        >
          <div
            class="selected"
          >
            <div
              aria-label=""
              class="m-tooltip-root tooltip-container"
              data-mui-internal-clone-element="true"
            >
              <span
                class="selected-text"
              >
                <span
                  class="placeholder"
                >
                  Select
                </span>
              </span>
            </div>
            <div
              class="chevron"
            >
              <svg
                class=""
                fill="#6F7880"
                height="10"
                viewBox="0 0 10 10"
                width="10"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m.5 3 4.1 4.1c.2.2.3.3.4.3s.2 0 .4-.2L9.6 3a.608.608 0 0 0 0-.8.608.608 0 0 0-.8 0L5 6.1 1.2 2.3a.483.483 0 0 0-.7 0 .483.483 0 0 0 0 .7Z"
                />
              </svg>
              <svg
                class="hide"
                fill="#6F7880"
                height="10"
                viewBox="0 0 10 10"
                width="10"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.6 6.5 5.4 2.3a.52.52 0 0 0-.4-.2.442.442 0 0 0-.4.2L.5 6.5a.385.385 0 0 0 0 .7.608.608 0 0 0 .8 0L5 3.4l3.8 3.8a.608.608 0 0 0 .8 0 .483.483 0 0 0 0-.7Z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
